File: components/kilix-95/provider_protocol.py
"""Dependency-free protocol-v1 adapter for the Kilix 95 entry point.

This module must stay safe to import before Pillow, ``kilix_sdk`` or any
provider storage is initialized.  Metadata and health probes are therefore
read-only and use only the Python standard library.
"""
from __future__ import annotations

import importlib.util
import json
import os
import re
import sys
from typing import Any


PROVIDER_ID = "kilix-95"
CONTRACT_VERSION = 1
EXIT_USAGE = 2
EXIT_INVALID_REQUEST = 3
EXIT_UNAVAILABLE = 4
_ROOT = os.path.dirname(os.path.abspath(__file__))
_SESSION_ID = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$")


def _version() -> str:
    try:
        with open(os.path.join(_ROOT, "VERSION"), encoding="utf-8") as handle:
            return handle.read().strip()
    except OSError:
        return "0.0.0-unknown"


def _unavailable(reason: str, detail: str) -> dict[str, object]:
    return {"available": False, "detail": detail, "reason": reason}


def _emit(document: dict[str, Any]) -> None:
    sys.stdout.write(
        json.dumps(
            document,
            ensure_ascii=True,
            separators=(",", ":"),
            sort_keys=True,
        )
        + "\n"
    )


def _description() -> dict[str, Any]:
    return {
        "capabilities": {
            "audio": True,
            "headless_screenshot": True,
            "keyboard": True,
            "launcher": True,
            "mouse": True,
            "reduced_motion": _unavailable(
                "not-implemented",
                "Kilix 95 does not expose a reduced-motion renderer.",
            ),
            "settings": _unavailable(
                "not-implemented",
                "Protocol configuration writes are not available in this adapter.",
            ),
        },
        "config_schema": "kilix.desktop.config.provider.v1",
        "contract_version": CONTRACT_VERSION,
        "display_modes": ["kitty-graphics"],
        "provider_id": PROVIDER_ID,
        "provider_version": _version(),
        "required_capabilities": ["keyboard"],
        "schema_version": 1,
    }


def _sdk_available() -> bool:
    candidates: list[str] = []
    if home := os.environ.get("KILIX_HOME"):
        candidates.append(os.path.join(home, "config", "kilix_sdk", "__init__.py"))
    if source := os.environ.get("GPU_TERMINAL_SOURCE_HOME"):
        candidates.append(
            os.path.join(source, "kilix", "config", "kilix_sdk", "__init__.py")
        )
    candidates.append(
        os.path.join(
            os.path.dirname(os.path.dirname(_ROOT)),
            "kilix",
            "config",
            "kilix_sdk",
            "__init__.py",
        )
    )
    return any(os.path.isfile(path) and os.access(path, os.R_OK) for path in candidates)


def _check() -> dict[str, Any]:
    source_ready = all(
        os.path.isfile(os.path.join(_ROOT, name))
        for name in ("theme.py", "shell.py", "wm.py")
    ) and os.path.isdir(os.path.join(_ROOT, "assets"))
    pillow_ready = importlib.util.find_spec("PIL") is not None
    sdk_ready = _sdk_available()
    checks = [
        {
            "id": "provider-files",
            "required": True,
            "status": "pass" if source_ready else "unavailable",
            "summary": (
                "Provider modules and assets are readable."
                if source_ready
                else "Provider modules or assets are missing."
            ),
        },
        {
            "id": "pillow",
            "required": True,
            "status": "pass" if pillow_ready else "unavailable",
            "summary": (
                "Pillow is importable."
                if pillow_ready
                else "Pillow is not importable."
            ),
            **(
                {}
                if pillow_ready
                else {"remediation": "Install the provider's locked Python dependencies."}
            ),
        },
        {
            "id": "kilix-sdk",
            "required": True,
            "status": "pass" if sdk_ready else "unavailable",
            "summary": (
                "The selected Kilix SDK is readable."
                if sdk_ready
                else "The selected Kilix SDK is not readable."
            ),
            **(
                {}
                if sdk_ready
                else {"remediation": "Launch through Kilix or set KILIX_HOME."}
            ),
        },
    ]
    ready = source_ready and pillow_ready and sdk_ready
    return {
        "checks": checks,
        "contract_version": CONTRACT_VERSION,
        "provider_id": PROVIDER_ID,
        "schema_version": 1,
        "status": "ready" if ready else "unavailable",
        "summary": (
            "Kilix 95 is ready."
            if ready
            else "Kilix 95 is missing a required runtime dependency."
        ),
    }


def _config_schema() -> dict[str, Any]:
    return {
        "$id": "https://schemas.kilix.org/desktop/config/kilix-95/v1",
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "additionalProperties": True,
        "properties": {},
        "type": "object",
        "x-kilix-contract-version": CONTRACT_VERSION,
        "x-kilix-provider-id": PROVIDER_ID,
    }


def _config_values() -> dict[str, Any]:
    return {
        "contract_version": CONTRACT_VERSION,
        "provider_id": PROVIDER_ID,
        "revision": 0,
        "schema_version": 1,
        "values": {},
    }


def _fail(message: str, status: int) -> None:
    print(f"kilix-95: {message}", file=sys.stderr)
    raise SystemExit(status)


def dispatch(argv: list[str]) -> list[str] | None:
    """Handle protocol commands or translate launch/screenshot to legacy argv.

    ``None`` means the argv is not a protocol command.  A returned list
    replaces the executable's argv; completed metadata endpoints exit here.
    """
    if not argv or argv[0] != "provider":
        return None

    if argv == ["provider", "describe", "--json"]:
        _emit(_description())
        raise SystemExit(0)
    if argv == ["provider", "check", "--json"]:
        _emit(_check())
        raise SystemExit(0)
    if argv == ["provider", "config", "schema", "--json"]:
        _emit(_config_schema())
        raise SystemExit(0)
    if argv in (
        ["provider", "config", "get", "--json"],
    ) or (
        len(argv) == 5
        and argv[:3] == ["provider", "config", "get"]
        and argv[4] == "--json"
        and bool(argv[3])
    ):
        _emit(_config_values())
        raise SystemExit(0)
    if len(argv) == 5 and argv[:3] == ["provider", "config", "set"]:
        _fail("protocol configuration writes are unavailable", EXIT_UNAVAILABLE)

    if argv[:2] == ["provider", "launch"]:
        if len(argv) == 2:
            return []
        if (
            len(argv) == 4
            and argv[2] == "--session-id"
            and _SESSION_ID.fullmatch(argv[3]) is not None
        ):
            os.environ["KILIX_DESKTOP_SESSION_ID"] = argv[3]
            return []
        if len(argv) == 4 and argv[2] == "--session-id":
            _fail("invalid provider session ID", EXIT_INVALID_REQUEST)
        _fail("usage: provider launch [--session-id ID]", EXIT_USAGE)

    if argv[:2] == ["provider", "screenshot"]:
        if len(argv) < 3 or not argv[2]:
            _fail("usage: provider screenshot OUTPUT [OPTIONS...]", EXIT_USAGE)
        os.environ["KILIX_PROVIDER_SCREENSHOT"] = "1"
        return ["--screenshot", argv[2], *argv[3:]]

    if argv[:2] == ["provider", "migrate"]:
        if (
            len(argv) not in (4, 5)
            or argv[2] != "--from"
            or not argv[3]
            or len(argv[3]) > 64
            or any(ord(character) < 0x20 or ord(character) == 0x7F for character in argv[3])
            or (len(argv) == 5 and argv[4] != "--dry-run")
        ):
            _fail("usage: provider migrate --from VERSION [--dry-run]", EXIT_USAGE)
        _fail("protocol persistence migration is unavailable", EXIT_UNAVAILABLE)

    _fail("unknown provider protocol command", EXIT_USAGE)
    raise AssertionError("unreachable")

File: components/kilix-land-desktop/src/provider_protocol.c
#include "provider_protocol.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

enum {
    KILIX_PROVIDER_EXIT_USAGE = 2,
    KILIX_PROVIDER_EXIT_INVALID_REQUEST = 3,
    KILIX_PROVIDER_EXIT_UNAVAILABLE = 4,
    KILIX_PROVIDER_EXIT_INTERNAL_ERROR = 70
};

static int emit_description(const struct kilix_provider_v1 *provider)
{
    return printf(
        "{\"capabilities\":%s,\"config_schema\":"
        "\"kilix.desktop.config.provider.v1\",\"contract_version\":1,"
        "\"display_modes\":%s,\"provider_id\":\"%s\","
        "\"provider_version\":\"%s\",\"required_capabilities\":%s,"
        "\"schema_version\":1}\n",
        provider->capabilities_json, provider->display_modes_json,
        provider->provider_id, provider->provider_version,
        provider->required_capabilities_json) < 0
               ? KILIX_PROVIDER_EXIT_INTERNAL_ERROR
               : 0;
}

static int emit_check(const struct kilix_provider_v1 *provider)
{
    const char *check_status = provider->check_ready ? "pass" : "unavailable";
    const char *status = provider->check_ready ? "ready" : "unavailable";
    const char *summary = provider->check_ready
                              ? provider->ready_summary
                              : provider->unavailable_summary;
    const char *check_summary = provider->check_ready
                                    ? provider->check_pass_summary
                                    : provider->check_unavailable_summary;
    return printf(
        "{\"checks\":[{\"id\":\"%s\",\"required\":true,"
        "\"status\":\"%s\",\"summary\":\"%s\"}],"
        "\"contract_version\":1,\"provider_id\":\"%s\","
        "\"schema_version\":1,\"status\":\"%s\",\"summary\":\"%s\"}\n",
        provider->check_id, check_status, check_summary,
        provider->provider_id, status, summary) < 0
               ? KILIX_PROVIDER_EXIT_INTERNAL_ERROR
               : 0;
}

static int emit_config_schema(const struct kilix_provider_v1 *provider)
{
    return printf(
        "{\"$id\":\"https://schemas.kilix.org/desktop/config/%s/v1\","
        "\"$schema\":\"https://json-schema.org/draft/2020-12/schema\","
        "\"additionalProperties\":true,\"properties\":{},\"type\":\"object\","
        "\"x-kilix-contract-version\":1,\"x-kilix-provider-id\":\"%s\"}\n",
        provider->provider_id, provider->provider_id) < 0
               ? KILIX_PROVIDER_EXIT_INTERNAL_ERROR
               : 0;
}

static int emit_config(const struct kilix_provider_v1 *provider)
{
    return printf(
        "{\"contract_version\":1,\"provider_id\":\"%s\",\"revision\":0,"
        "\"schema_version\":1,\"values\":{}}\n",
        provider->provider_id) < 0
               ? KILIX_PROVIDER_EXIT_INTERNAL_ERROR
               : 0;
}

static int ascii_alphanumeric(unsigned char character)
{
    return (character >= (unsigned char)'A' &&
            character <= (unsigned char)'Z') ||
           (character >= (unsigned char)'a' &&
            character <= (unsigned char)'z') ||
           (character >= (unsigned char)'0' &&
            character <= (unsigned char)'9');
}

static int valid_session_id(const char *value)
{
    size_t index;
    size_t length;
    if (value == NULL) return 0;
    length = strlen(value);
    if (length == 0u || length > 128u || !ascii_alphanumeric((unsigned char)value[0]))
        return 0;
    for (index = 1u; index < length; ++index) {
        unsigned char character = (unsigned char)value[index];
        if (!ascii_alphanumeric(character) && character != (unsigned char)'.' &&
            character != (unsigned char)'_' && character != (unsigned char)'-')
            return 0;
    }
    return 1;
}

static int valid_version(const char *value)
{
    size_t index;
    size_t length;
    if (value == NULL) return 0;
    length = strlen(value);
    if (length == 0u || length > 64u) return 0;
    for (index = 0u; index < length; ++index) {
        unsigned char character = (unsigned char)value[index];
        if (character < 0x20u || character == 0x7fu) return 0;
    }
    return 1;
}

static int unavailable(const char *provider_id, const char *feature)
{
    (void)fprintf(stderr, "%s: %s is unavailable\n", provider_id, feature);
    return KILIX_PROVIDER_EXIT_UNAVAILABLE;
}

int kilix_provider_v1_dispatch(int argc, char **argv,
                               const struct kilix_provider_v1 *provider)
{
    if (argc < 2 || strcmp(argv[1], "provider") != 0)
        return KILIX_PROVIDER_NOT_HANDLED;
    if (provider == NULL) return KILIX_PROVIDER_EXIT_INTERNAL_ERROR;

    if (argc == 4 && strcmp(argv[2], "describe") == 0 &&
        strcmp(argv[3], "--json") == 0)
        return emit_description(provider);
    if (argc == 4 && strcmp(argv[2], "check") == 0 &&
        strcmp(argv[3], "--json") == 0)
        return emit_check(provider);
    if (argc == 5 && strcmp(argv[2], "config") == 0 &&
        strcmp(argv[3], "schema") == 0 && strcmp(argv[4], "--json") == 0)
        return emit_config_schema(provider);
    if ((argc == 5 || argc == 6) && strcmp(argv[2], "config") == 0 &&
        strcmp(argv[3], "get") == 0 &&
        strcmp(argv[argc - 1], "--json") == 0 &&
        (argc == 5 || argv[4][0] != '\0'))
        return emit_config(provider);
    if (argc == 6 && strcmp(argv[2], "config") == 0 &&
        strcmp(argv[3], "set") == 0)
        return unavailable(provider->provider_id,
                           "protocol configuration writes");

    if (argc >= 3 && strcmp(argv[2], "launch") == 0) {
        if (argc == 3) return KILIX_PROVIDER_LAUNCH;
        if (argc == 5 && strcmp(argv[3], "--session-id") == 0) {
            if (!valid_session_id(argv[4])) {
                (void)fprintf(stderr, "%s: invalid provider session ID\n",
                              provider->provider_id);
                return KILIX_PROVIDER_EXIT_INVALID_REQUEST;
            }
            if (setenv("KILIX_DESKTOP_SESSION_ID", argv[4], 1) != 0)
                return KILIX_PROVIDER_EXIT_INTERNAL_ERROR;
            return KILIX_PROVIDER_LAUNCH;
        }
        (void)fprintf(stderr,
                      "%s: usage: provider launch [--session-id ID]\n",
                      provider->provider_id);
        return KILIX_PROVIDER_EXIT_USAGE;
    }

    if (argc >= 3 && strcmp(argv[2], "screenshot") == 0) {
        if (argc < 4 || argv[3][0] == '\0') {
            (void)fprintf(stderr,
                          "%s: usage: provider screenshot OUTPUT [OPTIONS...]\n",
                          provider->provider_id);
            return KILIX_PROVIDER_EXIT_USAGE;
        }
        if (!provider->screenshot_available)
            return unavailable(provider->provider_id, "headless screenshot");
        return KILIX_PROVIDER_SCREENSHOT;
    }

    if (argc >= 3 && strcmp(argv[2], "migrate") == 0) {
        if ((argc != 5 && argc != 6) || strcmp(argv[3], "--from") != 0 ||
            (argc == 6 && strcmp(argv[5], "--dry-run") != 0)) {
            (void)fprintf(stderr,
                          "%s: usage: provider migrate --from VERSION [--dry-run]\n",
                          provider->provider_id);
            return KILIX_PROVIDER_EXIT_USAGE;
        }
        if (!valid_version(argv[4])) {
            (void)fprintf(stderr, "%s: invalid migration source version\n",
                          provider->provider_id);
            return KILIX_PROVIDER_EXIT_INVALID_REQUEST;
        }
        return unavailable(provider->provider_id,
                           "protocol persistence migration");
    }

    (void)fprintf(stderr, "%s: unknown provider protocol command\n",
                  provider->provider_id);
    return KILIX_PROVIDER_EXIT_USAGE;
}
